### PR TITLE
[IMP] crm: improve partner opportunity count performance

### PR DESCRIPTION
**Pull request by yajo**
In v12, the `child_of` operator is quite heavy on the SQL level. In a database with many partners, this becomes slow. And the `opportunity_count`non-stored computed field appears on kanban views of `res.partner`, which makes performance worse.

Besides, instead of group-searching records, this was issuing one of such unoptimized queries *per record*. This is a really bad bottleneck.

By switching the domain to search for `id` and `commercial_partner_id`, and grouping results, we get the same results but with less queries.

In one production database, this reduces the load of kanban view of `res.partner` from 7s to 2.5s.

@Tecnativa TT30390

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/crm/models/res_partner.py [hunk 1]
```diff
@@ -38,9 +38,27 @@ def default_get(self, fields):
 
     @api.multi
     def _compute_opportunity_count(self):
+        result_list = self.env['crm.lead'].read_group(
+            [
+                ('type', '=', 'opportunity'),
+                '|',
+                ('partner_id.commercial_partner_id', 'in', self.ids),
```
Comment (pedrobaeza) [inline addons/crm/models/res_partner.py:45]: `child_of` is needed, as you can have more than 2 levels.

**Comment by pedrobaeza** [comment]
@yelizariev this one can fit in your performance work that you are doing, so would be glad if you review it and push for it's inclusion. It's a real bottleneck we have.

**Comment by yelizariev** [comment]
Is it for v12 only?

Did you try implementation from v14 ?

https://github.com/odoo/odoo/blame/c63af09774414a004ef34644b4f38b4fdb1d12bd/addons/crm/models/res_partner.py#L39-L55

File: addons/crm/models/res_partner.py (hunk 1 above)
Comment (Yajo) [reply addons/crm/models/res_partner.py:45]: Solved, cross-posting answer from https://github.com/odoo/odoo/pull/71950#issuecomment-858553512

**Comment by Yajo** [comment]
> Is it for v12 only?

AFAICS it would affect still v13.

> Did you try implementation from v14 ?

Nope. Seems quite similar, although that one wouldn't serve here, where we still don't have the in-cache ORM, so I don't think it's worth backporting. I'd have to change almost all lines anyways... 🤷🏼‍♂️ 

However, for fixing v13, it'd be better to backport the v14 fix indeed.

**Comment by yelizariev** [comment]
@fw-bot ignore

(or saas-12.3 ?)

**Comment by C3POdoo** [comment]
Dear @yajo,

Thank you for your contribution but the version 12.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._

**Pull request closed without merging**